### PR TITLE
Expand entity

### LINKED ISSUE #152: expand entity in jsonschema schema
https://github.com/citation-file-format/citation-file-format/blob/04ec97816b5f069c205885642027eb50a5a17135/schema.json#L668-L671

### PR DESCRIPTION
**Pull request by sdruskat**
**Related issues**

Refs: fixes #152.

**Describe the changes made in this pull request**

- Create definitions for objects duplicated over person and entity
- Add all entity fields
- Fixes #152 
- According to https://json-schema.org/understanding-json-schema/reference/combining.html#id5, proper fixing of #103 in JSON Schema cannot be done with draft 7, hints to fixing in draft 8 though
-  Fix bad tel definition, add tel ref
- Add test data for entity

**Instructions to review the pull request**

**Q:** Externalizing definitions for reused simple string properties like this is one option, another one would be creating a `"_simple-string"` definition to reference which would at least help catch errors on redefinition of fields at some point (e.g., if a stricter definition of any of the reused fields is required). Another option would be to simply leave as "string", which could lead to not catching reuse across person/entity and only redefining one instance. Thoughts?

```bash
git clone this repo
git checkout this branch
python3 -m venv env
source env/bin/activate
pip install --upgrade pip setuptools wheel
pip install -r requirements.txt
python3 schema_poc.py -d data.yml -s schema.json  # <- should be quiet
```

### CODE REVIEW

File: schema.json [hunk 1]
```diff
@@ -979,9 +993,69 @@
...
+                    "$ref": "#/definitions/url"
+                },
+                "date-start": {
+                    "$ref": "#/definitions/date"
+                },
+                "date-end": {
+                    "$ref": "#/definitions/date"
+                },
+                "location": {
+                    "type": "string"
```
Comment (jspaaks) [inline schema.json:1037]: Somewhere around here there should be a `required` with `name` in it methinks

File: schema.json [hunk 2]
```diff
@@ -1351,4 +1465,4 @@
             "pattern": "^(?:(?:https?|ftp)://)(?:\\S+(?::\\S*)?@)?(?:(?!(?:10|127)(?:\\.\\d{1,3}){3})(?!(?:169\\.254|192\\.168)(?:\\.\\d{1,3}){2})(?!172\\.(?:1[6-9]|2\\d|3[0-1])(?:\\.\\d{1,3}){2})(?:[1-9]\\d?|1\\d\\d|2[01]\\d|22[0-3])(?:\\.(?:1?\\d{1,2}|2[0-4]\\d|25[0-5])){2}(?:\\.(?:[1-9]\\d?|1\\d\\d|2[0-4]\\d|25[0-4]))|(?:(?:[a-z\\u00a1-\\uffff0-9]-?)*[a-z\\u00a1-\\uffff0-9]+)(?:\\.(?:[a-z\\u00a1-\\uffff0-9]-?)*[a-z\\u00a1-\\uffff0-9]+)*(?:\\.(?:[a-z\\u00a1-\\uffff]{2,})))(?::\\d{2,5})?(?:/\\S*)?$"
         }
     }
-}
+}
```
Comment (jspaaks) [inline schema.json:1468]:
```suggestion
}

```

**Review by jspaaks — CHANGES_REQUESTED** [review]
Great! Just one small thing, otherwise good to go

**Comment by jspaaks** [comment]
> Externalizing definitions for reused simple string properties like this is one option, another one would be creating a "_simple-string" definition to reference which would at least help catch errors on redefinition of fields at some point (e.g., if a stricter definition of any of the reused fields is required). Another option would be to simply leave as "string", which could lead to not catching reuse across person/entity and only redefining one instance. Thoughts?

I would like to stay away from the `_simple_string` option. Ideally, I would also like to stay away from adding definitions for primitive types, but it looks like we currently need it, so that's OK for now I guess. I think it could be solved (in the future, not in this PR) by having three sets of keys `uniqueToEntity`, `uniqueToPerson`, and `shared`, and then the definition could be

```json
"entity": {
    "allOf": [ 
        {"$ref": "#/definitions/uniqueToEntity"},
        {"$ref": "#/definitions/shared"}
    ]
},
"person": {
    "allOf": [ 
        {"$ref": "#/definitions/uniqueToPerson"},
        {"$ref": "#/definitions/shared"}
    ]
}
```

Needs looking into.

File: schema.json (hunk 1 above)
Comment (sdruskat) [reply schema.json:1037]: Fixed in a162d1eb9b27e721b99fa73a23a4e4c26368acd1.

**Comment by sdruskat** [comment]
Thanks for the review. Made `name` required in https://github.com/citation-file-format/citation-file-format/pull/155#issuecomment-842456834, and added the newline.

Also spawned issue #164.

**Review by jspaaks — APPROVED** [review]
All good

**Comment by sdruskat** [comment]
Cheers!

**Pull request merged**